File: examples/primitives/transaction/verify-recover-sender.zig
// Transaction Signature Verification and Sender Recovery Example
//
// Demonstrates:
// - Checking if transaction is signed
// - Understanding signature components (r, s, yParity/v)
// - Converting between v and yParity
// - Signature malleability protection
// - Different signature formats across transaction types

const std = @import("std");
const primitives = @import("primitives");

const Transaction = primitives.Transaction;
const LegacyTransaction = Transaction.LegacyTransaction;
const EIP1559Transaction = Transaction.EIP1559Transaction;
const EIP7702Transaction = Transaction.EIP7702Transaction;
const Address = primitives.Address.Address;
const Authorization = Transaction.Authorization;

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    std.debug.print("\n=== Transaction Signature Verification ===\n\n", .{});

    // Example 1: Check if transaction is signed
    std.debug.print("1. Checking Transaction Signature Status\n", .{});
    std.debug.print("-" ** 50 ++ "\n", .{});

    const recipient = try Address.fromHex("0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e");

    const unsigned_r = [_]u8{0} ** 32; // All zeros = unsigned
    const unsigned_s = [_]u8{0} ** 32;

    var signed_r = [_]u8{0} ** 32;
    var signed_s = [_]u8{0} ** 32;
    @memset(&signed_r, 0x12);
    @memset(&signed_s, 0xfe);

    std.debug.print("Unsigned transaction:\n", .{});
    std.debug.print("  r: 0x{X:0>64}\n", .{std.fmt.fmtSliceHexLower(&unsigned_r)});
    std.debug.print("  s: 0x{X:0>64}\n", .{std.fmt.fmtSliceHexLower(&unsigned_s)});
    std.debug.print("  isSigned: {}\n", .{!isAllZeros(&unsigned_r) and !isAllZeros(&unsigned_s)});
    std.debug.print("\n", .{});

    std.debug.print("Signed transaction:\n", .{});
    std.debug.print("  r: 0x{X:0>64}\n", .{std.fmt.fmtSliceHexLower(&signed_r)});
    std.debug.print("  s: 0x{X:0>64}\n", .{std.fmt.fmtSliceHexLower(&signed_s)});
    std.debug.print("  isSigned: {}\n", .{!isAllZeros(&signed_r) and !isAllZeros(&signed_s)});
    std.debug.print("\n", .{});

    // Example 2: Signature components for different transaction types
    std.debug.print("2. Signature Components by Transaction Type\n", .{});
    std.debug.print("-" ** 50 ++ "\n", .{});

    const legacy_tx = LegacyTransaction{
        .nonce = 0,
        .gas_price = 20_000_000_000,
        .gas_limit = 21000,
        .to = recipient,
        .value = 1_000_000_000_000_000_000,
        .data = &[_]u8{},
        .v = 37, // EIP-155: chainId=1, yParity=0
        .r = signed_r,
        .s = signed_s,
    };

    std.debug.print("Legacy Transaction:\n", .{});
    std.debug.print("  Signature format: v/r/s\n", .{});
    std.debug.print("  v: {} (includes chain ID)\n", .{legacy_tx.v});
    std.debug.print("  r: 0x{s}...\n", .{std.fmt.fmtSliceHexLower(legacy_tx.r[0..8])});
    std.debug.print("  s: 0x{s}...\n", .{std.fmt.fmtSliceHexLower(legacy_tx.s[0..8])});
    std.debug.print("\n", .{});

    const eip1559_tx = EIP1559Transaction{
        .chain_id = 1,
        .nonce = 0,
        .max_priority_fee_per_gas = 1_000_000_000,
        .max_fee_per_gas = 20_000_000_000,
        .gas_limit = 21000,
        .to = recipient,
        .value = 1_000_000_000_000_000_000,
        .data = &[_]u8{},
        .access_list = &[_]Transaction.AccessListItem{},
        .y_parity = 0,
        .r = signed_r,
        .s = signed_s,
    };

    std.debug.print("EIP-1559 Transaction:\n", .{});
    std.debug.print("  Signature format: yParity/r/s\n", .{});
    std.debug.print("  yParity: {} (0 or 1 only)\n", .{eip1559_tx.y_parity});
    std.debug.print("  r: 0x{s}...\n", .{std.fmt.fmtSliceHexLower(eip1559_tx.r[0..8])});
    std.debug.print("  s: 0x{s}...\n", .{std.fmt.fmtSliceHexLower(eip1559_tx.s[0..8])});
    std.debug.print("\n", .{});

    // Example 3: Converting between v and yParity
    std.debug.print("3. Converting Between v and yParity\n", .{});
    std.debug.print("-" ** 50 ++ "\n", .{});

    const chain_id: u64 = 1;
    const y_parity: u8 = 0;
    const v = calculateVFromYParity(y_parity, chain_id);
    const extracted_y_parity = extractYParityFromV(v, chain_id);

    std.debug.print("Chain ID: {}\n", .{chain_id});
    std.debug.print("yParity: {}\n", .{y_parity});
    std.debug.print("Calculated v: {}\n", .{v});
    std.debug.print("Extracted yParity: {}\n", .{extracted_y_parity});
    std.debug.print("\n", .{});

    std.debug.print("Conversion formulas:\n", .{});
    std.debug.print("  yParity → v: v = chain_id * 2 + 35 + y_parity\n", .{});
    std.debug.print("  v → yParity: y_parity = (v - 35 - chain_id * 2)\n", .{});
    std.debug.print("\n", .{});

    // Example 4: Signature malleability protection
    std.debug.print("4. Signature Malleability Protection\n", .{});
    std.debug.print("-" ** 50 ++ "\n", .{});

    // secp256k1 curve order
    const SECP256K1_N = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141;
    const SECP256K1_N_DIV_2 = SECP256K1_N / 2;

    std.debug.print("Ethereum requires s ≤ n/2 to prevent signature malleability.\n", .{});
    std.debug.print("\n", .{});
    std.debug.print("secp256k1 curve order (n):\n", .{});
    std.debug.print("  0x{X}\n", .{SECP256K1_N});
    std.debug.print("\n", .{});
    std.debug.print("Maximum allowed s (n/2):\n", .{});
    std.debug.print("  0x{X}\n", .{SECP256K1_N_DIV_2});
    std.debug.print("\n", .{});
    std.debug.print("Why? For every (r, s), there exists (r, n - s) that validates.\n", .{});
    std.debug.print("Requiring s ≤ n/2 makes signatures canonical (unique).\n", .{});
    std.debug.print("\n", .{});

    // Example 5: EIP-7702 dual signatures
    std.debug.print("5. EIP-7702: Authorization vs Transaction Signature\n", .{});
    std.debug.print("-" ** 50 ++ "\n", .{});

    var auth_r = [_]u8{0} ** 32;
    var auth_s = [_]u8{0} ** 32;
    var tx_r = [_]u8{0} ** 32;
    var tx_s = [_]u8{0} ** 32;

    @memset(&auth_r, 0xaa);
    @memset(&auth_s, 0xbb);
    @memset(&tx_r, 0xcc);
    @memset(&tx_s, 0xdd);

    const contract_addr = try Address.fromHex("0x1234567890123456789012345678901234567890");

    var authorization_list = try allocator.alloc(Authorization, 1);
    defer allocator.free(authorization_list);

    authorization_list[0] = Authorization{
        .chain_id = 1,
        .address = contract_addr,
        .nonce = 0,
        .y_parity = 0,
        .r = auth_r,
        .s = auth_s,
    };

    std.debug.print("EIP-7702 has TWO signatures:\n\n", .{});
    std.debug.print("1. Authorization signature (signed by delegating EOA):\n", .{});
    std.debug.print("   yParity: {}\n", .{authorization_list[0].y_parity});
    std.debug.print("   r: 0x{s}...\n", .{std.fmt.fmtSliceHexLower(authorization_list[0].r[0..8])});
    std.debug.print("   s: 0x{s}...\n", .{std.fmt.fmtSliceHexLower(authorization_list[0].s[0..8])});
    std.debug.print("\n", .{});
    std.debug.print("2. Transaction signature (signed by transaction sender):\n", .{});
    std.debug.print("   yParity: 1\n", .{});
    std.debug.print("   r: 0x{s}...\n", .{std.fmt.fmtSliceHexLower(tx_r[0..8])});
    std.debug.print("   s: 0x{s}...\n", .{std.fmt.fmtSliceHexLower(tx_s[0..8])});
    std.debug.print("\n", .{});
    std.debug.print("Both must be valid for transaction to execute!\n", .{});
    std.debug.print("\n", .{});

    // Example 6: Sender recovery process
    std.debug.print("6. Sender Recovery Process\n", .{});
    std.debug.print("-" ** 50 ++ "\n", .{});

    std.debug.print("Steps to recover sender address:\n", .{});
    std.debug.print("  1. Get signing hash (keccak256 of transaction data)\n", .{});
    std.debug.print("  2. Recover public key using secp256k1.recover(hash, r, s, yParity)\n", .{});
    std.debug.print("  3. Hash public key with keccak256\n", .{});
    std.debug.print("  4. Take last 20 bytes as address\n", .{});
    std.debug.print("\n", .{});

    std.debug.print("Note: Actual implementation requires secp256k1 crypto library.\n", .{});
    std.debug.print("\n", .{});

    // Example 7: Practical use cases
    std.debug.print("7. Practical Use Cases\n", .{});
    std.debug.print("-" ** 50 ++ "\n", .{});

    std.debug.print("Transaction Pool Validation:\n", .{});
    std.debug.print("  1. Verify signature is valid\n", .{});
    std.debug.print("  2. Recover sender address\n", .{});
    std.debug.print("  3. Check sender has sufficient balance\n", .{});
    std.debug.print("  4. Check sender nonce matches transaction nonce\n", .{});
    std.debug.print("\n", .{});

    std.debug.print("Replay Attack Protection:\n", .{});
    std.debug.print("  1. Verify chain ID in transaction\n", .{});
    std.debug.print("  2. Ensure chain ID matches current network\n", .{});
    std.debug.print("  3. Check signature was created for this chain\n", .{});
    std.debug.print("\n", .{});

    std.debug.print("Authorization Checking:\n", .{});
    std.debug.print("  1. Recover sender from signature\n", .{});
    std.debug.print("  2. Compare with expected sender\n", .{});
    std.debug.print("  3. Reject if mismatch\n", .{});

    std.debug.print("\n=== Example Complete ===\n\n", .{});
}

// Helper functions

fn isAllZeros(bytes: []const u8) bool {
    for (bytes) |b| {
        if (b != 0) return false;
    }
    return true;
}

fn calculateVFromYParity(y_parity: u8, chain_id: u64) u64 {
    // v = chain_id * 2 + 35 + y_parity
    return chain_id * 2 + 35 + y_parity;
}

fn extractYParityFromV(v: u64, chain_id: u64) u8 {
    // For EIP-155: v = chain_id * 2 + 35 + y_parity
    // y_parity = (v - 35 - chain_id * 2)
    if (v == 27 or v == 28) {
        // Pre-EIP-155
        return @intCast(v - 27);
    }
    return @intCast(v - 35 - chain_id * 2);
}

File: examples/primitives/transaction/verify-recover-sender.ts
/**
 * Transaction Signature Verification and Sender Recovery Example
 *
 * Demonstrates:
 * - Recovering sender address from transaction signature
 * - Verifying transaction signatures
 * - Checking if transaction is signed
 * - Understanding signature components (r, s, yParity/v)
 * - Batch signature verification
 */

import * as Transaction from '../../../src/primitives/Transaction/index.js';
import * as Address from '../../../src/primitives/Address/index.js';
import * as Hex from '../../../src/primitives/Hex/index.js';

console.log('=== Transaction Signature Verification and Sender Recovery ===\n');

// Example 1: Check if transaction is signed
console.log('1. Checking Transaction Signature Status:');
console.log('-'.repeat(50));

const unsignedTx: Transaction.EIP1559 = {
  type: Transaction.Type.EIP1559,
  chainId: 1n,
  nonce: 0n,
  maxPriorityFeePerGas: 1_000_000_000n,
  maxFeePerGas: 20_000_000_000n,
  gasLimit: 21000n,
  to: Address.from('0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e'),
  value: 1_000_000_000_000_000_000n,
  data: new Uint8Array(),
  accessList: [],
  yParity: 0,
  r: new Uint8Array(32),  // All zeros = unsigned
  s: new Uint8Array(32),  // All zeros = unsigned
};

const signedTx: Transaction.EIP1559 = {
  ...unsignedTx,
  r: Hex.toBytes('0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef'),
  s: Hex.toBytes('0xfedcba0987654321fedcba0987654321fedcba0987654321fedcba0987654321'),
};

console.log('Unsigned transaction:');
console.log('  r:', Hex.fromBytes(unsignedTx.r));
console.log('  s:', Hex.fromBytes(unsignedTx.s));
console.log('  isSigned:', Transaction.isSigned(unsignedTx));
console.log();

console.log('Signed transaction:');
console.log('  r:', Hex.fromBytes(signedTx.r));
console.log('  s:', Hex.fromBytes(signedTx.s));
console.log('  isSigned:', Transaction.isSigned(signedTx));
console.log();

// Example 2: Asserting transaction is signed
console.log('2. Asserting Signature Presence:');
console.log('-'.repeat(50));

try {
  Transaction.assertSigned(unsignedTx);
  console.log('✓ Unsigned transaction passed assertion (should not happen!)');
} catch (error) {
  console.log('✗ Unsigned transaction failed assertion (expected):');
  console.log('  Error:', (error as Error).message);
}

try {
  Transaction.assertSigned(signedTx);
  console.log('✓ Signed transaction passed assertion (expected)');
} catch (error) {
  console.log('✗ Signed transaction failed assertion (should not happen!)');
}
console.log();

// Example 3: Recovering sender address (conceptual - actual recovery requires valid signature)
console.log('3. Sender Address Recovery:');
console.log('-'.repeat(50));
console.log('Note: This example shows the API. Real signatures require secp256k1.\n');

// In real usage, you'd have a properly signed transaction
const exampleSignedTx: Transaction.EIP1559 = {
  type: Transaction.Type.EIP1559,
  chainId: 1n,
  nonce: 0n,
  maxPriorityFeePerGas: 1_000_000_000n,
  maxFeePerGas: 20_000_000_000n,
  gasLimit: 21000n,
  to: Address.from('0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e'),
  value: 1_000_000_000_000_000_000n,
  data: new Uint8Array(),
  accessList: [],
  yParity: 0,
  r: Hex.toBytes('0x' + '12'.repeat(32)),
  s: Hex.toBytes('0x' + 'fe'.repeat(32)),
};

console.log('Transaction details:');
console.log('  To:', Address.toHex(exampleSignedTx.to!));
console.log('  Value:', exampleSignedTx.value / 1_000_000_000_000_000_000n, 'ETH');
console.log('  yParity:', exampleSignedTx.yParity);
console.log('  r:', Hex.fromBytes(exampleSignedTx.r).slice(0, 20) + '...');
console.log('  s:', Hex.fromBytes(exampleSignedTx.s).slice(0, 20) + '...');
console.log();

// Note: In production, getSender() would perform ECDSA recovery
console.log('Sender recovery process:');
console.log('  1. Get signing hash (keccak256 of transaction data)');
console.log('  2. Recover public key using secp256k1.recover(hash, r, s, yParity)');
console.log('  3. Hash public key with keccak256');
console.log('  4. Take last 20 bytes as address');
console.log();

// Example 4: Signature components for different transaction types
console.log('4. Signature Components by Transaction Type:');
console.log('-'.repeat(50));

const legacyTx: Transaction.Legacy = {
  type: Transaction.Type.Legacy,
  nonce: 0n,
  gasPrice: 20_000_000_000n,
  gasLimit: 21000n,
  to: Address.from('0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e'),
  value: 1_000_000_000_000_000_000n,
  data: new Uint8Array(),
  v: 37n,  // EIP-155: chainId=1, yParity=0
  r: Hex.toBytes('0x' + '12'.repeat(32)),
  s: Hex.toBytes('0x' + 'fe'.repeat(32)),
};

console.log('Legacy Transaction:');
console.log('  Signature format: v/r/s');
console.log('  v:', legacyTx.v, '(includes chain ID)');
console.log('  r:', Hex.fromBytes(legacyTx.r).slice(0, 20) + '...');
console.log('  s:', Hex.fromBytes(legacyTx.s).slice(0, 20) + '...');
console.log();

console.log('EIP-1559 Transaction:');
console.log('  Signature format: yParity/r/s');
console.log('  yParity:', exampleSignedTx.yParity, '(0 or 1 only)');
console.log('  r:', Hex.fromBytes(exampleSignedTx.r).slice(0, 20) + '...');
console.log('  s:', Hex.fromBytes(exampleSignedTx.s).slice(0, 20) + '...');
console.log();

// Example 5: Converting between v and yParity
console.log('5. Converting Between v and yParity:');
console.log('-'.repeat(50));

function extractYParityFromV(v: bigint, chainId: bigint): number {
  // For EIP-155: v = chainId * 2 + 35 + yParity
  // yParity = (v - 35 - chainId * 2)
  if (v === 27n || v === 28n) {
    // Pre-EIP-155
    return Number(v - 27n);
  }
  return Number((v - 35n - chainId * 2n));
}

function calculateVFromYParity(yParity: number, chainId: bigint): bigint {
  // v = chainId * 2 + 35 + yParity
  return chainId * 2n + 35n + BigInt(yParity);
}

const chainId = 1n;
const yParity = 0;
const v = calculateVFromYParity(yParity, chainId);
const extractedYParity = extractYParityFromV(v, chainId);

console.log(`Chain ID: ${chainId}`);
console.log(`yParity: ${yParity}`);
console.log(`Calculated v: ${v}`);
console.log(`Extracted yParity: ${extractedYParity}`);
console.log();

// Example 6: Batch signature verification pattern
console.log('6. Batch Signature Verification:');
console.log('-'.repeat(50));

const transactions: Transaction.EIP1559[] = [
  exampleSignedTx,
  { ...exampleSignedTx, nonce: 1n },
  { ...exampleSignedTx, nonce: 2n },
];

console.log(`Verifying ${transactions.length} transactions:\n`);

let validCount = 0;
let invalidCount = 0;

for (let i = 0; i < transactions.length; i++) {
  const tx = transactions[i];
  const isSigned = Transaction.isSigned(tx);

  console.log(`Transaction ${i}:`);
  console.log(`  Nonce: ${tx.nonce}`);
  console.log(`  Signed: ${isSigned}`);

  if (isSigned) {
    validCount++;
  } else {
    invalidCount++;
  }
}

console.log();
console.log(`Summary:`);
console.log(`  Total: ${transactions.length}`);
console.log(`  Signed: ${validCount}`);
console.log(`  Unsigned: ${invalidCount}`);
console.log();

// Example 7: Signature malleability protection
console.log('7. Signature Malleability Protection:');
console.log('-'.repeat(50));

// secp256k1 curve order
const SECP256K1_N = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141n;
const SECP256K1_N_DIV_2 = SECP256K1_N / 2n;

console.log('Ethereum requires s ≤ n/2 to prevent signature malleability.');
console.log();
console.log('secp256k1 curve order (n):');
console.log(`  0x${SECP256K1_N.toString(16)}`);
console.log();
console.log('Maximum allowed s (n/2):');
console.log(`  0x${SECP256K1_N_DIV_2.toString(16)}`);
console.log();
console.log('Why? For every (r, s), there exists (r, n - s) that validates.');
console.log('Requiring s ≤ n/2 makes signatures canonical (unique).');
console.log();

// Example 8: Authorization vs transaction signature
console.log('8. EIP-7702: Authorization vs Transaction Signature:');
console.log('-'.repeat(50));

const eip7702Tx: Transaction.EIP7702 = {
  type: Transaction.Type.EIP7702,
  chainId: 1n,
  nonce: 0n,
  maxPriorityFeePerGas: 1_000_000_000n,
  maxFeePerGas: 20_000_000_000n,
  gasLimit: 100_000n,
  to: Address.from('0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e'),
  value: 0n,
  data: new Uint8Array(),
  accessList: [],
  authorizationList: [
    {
      chainId: 1n,
      address: Address.from('0x1234567890123456789012345678901234567890'),
      nonce: 0n,
      yParity: 0,
      r: Hex.toBytes('0x' + 'aa'.repeat(32)),
      s: Hex.toBytes('0x' + 'bb'.repeat(32)),
    },
  ],
  yParity: 1,
  r: Hex.toBytes('0x' + 'cc'.repeat(32)),
  s: Hex.toBytes('0x' + 'dd'.repeat(32)),
};

console.log('EIP-7702 has TWO signatures:\n');
console.log('1. Authorization signature (signed by delegating EOA):');
console.log('   yParity:', eip7702Tx.authorizationList[0].yParity);
console.log('   r:', Hex.fromBytes(eip7702Tx.authorizationList[0].r).slice(0, 20) + '...');
console.log('   s:', Hex.fromBytes(eip7702Tx.authorizationList[0].s).slice(0, 20) + '...');
console.log();
console.log('2. Transaction signature (signed by transaction sender):');
console.log('   yParity:', eip7702Tx.yParity);
console.log('   r:', Hex.fromBytes(eip7702Tx.r).slice(0, 20) + '...');
console.log('   s:', Hex.fromBytes(eip7702Tx.s).slice(0, 20) + '...');
console.log();
console.log('Both must be valid for transaction to execute!');
console.log();

// Example 9: Practical sender recovery use case
console.log('9. Practical Use Cases:');
console.log('-'.repeat(50));

console.log('Transaction Pool Validation:');
console.log('  1. Verify signature is valid');
console.log('  2. Recover sender address');
console.log('  3. Check sender has sufficient balance');
console.log('  4. Check sender nonce matches transaction nonce');
console.log();

console.log('Replay Attack Protection:');
console.log('  1. Verify chain ID in transaction');
console.log('  2. Ensure chain ID matches current network');
console.log('  3. Check signature was created for this chain');
console.log();

console.log('Authorization Checking:');
console.log('  1. Recover sender from signature');
console.log('  2. Compare with expected sender');
console.log('  3. Reject if mismatch');
console.log();
